### ISSUE TITLE
New snapshot mechanism

### ISSUE THREAD

**Issue by Flrande** [toeverything/blocksuite#3583]
Snapshot is a serialization format for data within blocksuite. It is required to support the conversion and migration between Snapshot JSON data and blocksuite runtime data.

```ts
export interface IBlockSnapshot {
  type: 'block';
  version: number;
  data: {
    // data in "meta" set with "sys" prefix
    meta: {
      id: string;
      flavour: string;
      children: string[];
    };
    // custom properties set with "prop" prefix
    [k: string]: unknown;
  };
}

export interface IPageSnapshot {
  type: 'page';
  version: number;
  data: {
    // data in "meta" will be stored in `workspace.meta.pages`
    meta: {
      id: string;
      title: string;
      createDate: number;
      tags: string[];
    };
    blocks: IBlockSnapshot[];
  };
}

export interface IWorkspaceSnapshot {
  type: 'workspace';
  data: {
    meta: {
      // the single source of truth of what blocks version is in the workspace
      blockVersions: {
        [k: string]: number;
      };
      pages: {
        [k: string]: unknown;
      }[];
      properties: {
        [k: string]: unknown;
      };
      workspaceVersion: number;
      pageVersion: number;
    };
    spaces: IPageSnapshot[];
  };
}
```

# Snapshot -> blocksuite runtime data

1. import block snapshot in page: `page.importBlockSnapshot(blockSnapshot[])`
  mirgrate each `blockSnapshot` based `blockVersions` in `page.workspace.meta`
    (`blockSnapshot.version` -> `page.workspace.meta.blockVersions[blockSnapshot.data.meta.flavour]`)

2. import page snapshot in workspace: `workspace.importPageSnapshot(pageSnapshot[])`
  2.1. mirgrate each `pageSnapshot` based `workspace.meta.pageVersion`
         (`pageSnapshot.version` -> `workspace.meta.pageVersion`)
  2.2. mirgrate each block based `blockVersions` in `workspace.meta`
         (use `1` for `blocks` in `pageSnapshot`)

3. init workspace with workspace snapshot:
  3.1. mirgrate workspace meta
         (`workspaceSnapshot.data.meta.workspaceVersion` -> version specified in runtime)
  3.2. mirgrate each pageSnapshot based `workspaceSnapshot.data.meta.pageVersion`
         (use `2` for `spaces` in `workspaceSnapshot`)

# What types of old data do we need to process?

1. Snapshot
2. The Y.Doc of the entire workspace 

# What methods do we need to implemented?

### `exportSnapshot` in Workspace and Page

A method is implemented separately on both Workspace and Page, which generates corresponding snapshots throughout, as shown in the above format.

### `exportBlockSnapshot` in Page

A method implemented on Page to export the corresponding block snapshot.

### Workspace Constructor

Allow the constructor of the workspace to support the restoration of a workspace from an existing workspace snapshot. This process involves migration operations.

### `importPageSnapshot` in Workspace

This is a method implemented on the workspace to import a page snapshot onto an existing workspace. During the import process, a **migration** operation will be executed.

### `importBlockSnapshot` in Page

This is a method implemented on the page to import a block snapshot onto an existing page. During the import process, a **migration** operation will be executed.

**Comment by Flrande** [comment]
This is still a draft, and there may be some scenarios that haven't been taken into account, such as preserving the relationships between pages in a workspace and references to blocks in pageA within pageB.

cc @Saul-Mirone @doodlewind @zzj3720

**Comment by Flrande** [comment]
Snapshot is closely related to Transformer, so the [migration part](https://github.com/toeverything/blocksuite/issues/3618) has been separated from this issue, and the development of snapshot has been postponed.

**Comment by doodlewind** [comment]
We have revamped the target and will track this in #3654, which is planned in next release. Thanks for the initial design effort!
